Update normalization notebook for tabula muris data (#457)
AlexsLemonade/training-modules: scRNA-seq/03-normalizing_scRNA.Rmd

```{r filepaths}
# main data directory
data_dir <- file.path("data", "tabula-muris")

# Filtered count matrix file from previous notebook
filtered_sce_file <- file.path(data_dir, "filtered", "filtered_sce.rds")

# Metadata file location
metadata_file <- file.path(data_dir, "TM_droplet_metadata.csv")

# Output directory for normalized data
norm_dir <- file.path(data_dir, "normalized")
if (!dir.exists(norm_dir)) {
  dir.create(norm_dir, recursive = TRUE)
}

# Output directory for markers
marker_dir <- file.path("analysis", "tabula-muris", "markers")
if (!dir.exists(marker_dir)) {
  dir.create(marker_dir, recursive = TRUE)
}
```

## Read in the filtered count matrix and metadata

```{r read_data}
bladder_sce <- readr::read_rds(filtered_sce_file)
sc_metadata <- readr::read_csv(metadata_file, guess_max = 10000)
```

### Adding more metadata to the SCE object

> In a previous notebook, we used some of the data from the Tabula Muris project analysis to tag our cells for visualization.
> Note that we would normally **NOT** have this information until later in the analysis pipeline!
> Nonetheless, adding it here will be useful for visualizing the results of our normalization (and demonstrating how one might add metadata to the SingleCellExperiment object).
> 

Reviewer: ```suggestion
Because this is a well-studied data set, we actually have some cell type information for this data set.

Note that we would normally **NOT** have this information until later in the analysis pipeline!
Nonetheless, adding it here will be useful for visualizing the results of our normalization (and demonstrating how one might add metadata to the SingleCellExperiment object).

```

Since we plan on taking out the cell type gene expression plots in the second notebook on filtering, we should probably change this description here.

Author: > I like the changes that you made here and thinks all looks good in terms of content. I added a few areas where I thought we could use some clarity, but other than that I didn't think there was much need for any major changes.
> 
> The one thing I did want to ask was about the lack of live typing in this notebook? It looks like there's only a handful of chunks that are live typing. I think we could probably increase that and use the smaller chunks as live to keep people engaged.

I agree there should probably be more live typing here; I don't think I was too focused on that when making the current edits. I think I planned to come back through with those changes a bit later, once I had all the content sorted & running properly.

I like your other edits too. I'm going to take a pass at wording tomorrow (and getting all the suggestions where they go, since github only lets you comment on lines near changes).

cell_metadata <- sc_metadata %>%
  # filter to just the sample we are working with
  dplyr::filter(channel == "10X_P4_3") %>%
  # extract 16 nt cell barcodes from the `cell` column
  dplyr::mutate(barcode = stringr::str_sub(cell, start= -16)) %>%
  # choose only the columns we want to add
  dplyr::select(c("barcode", "cell_ontology_class", "free_annotation"))

# Join the tables together, using `left_join()` to preserve all rows in cell_info
cell_info <- cell_info %>%
  dplyr::left_join(cell_metadata)
```

Check that the sample info accession ids are still the same as the columns of our data

```{r check_sampleinfo, live = TRUE}
all.equal(cell_info$barcode, colnames(bladder_sce))
```

Now we can add that data back to the SCE object.
For unclear reasons, we have to convert our table to a `DataFrame` object in order for this to work.
Just to keep things confusing, a `DataFrame` is not the same as a `data.frame` that we have been using throughout. 
We also need to be sure to include the `row.names` argument to keep those properly attached.

Note that this will replace all of the previous column (cell) metadata, which is why we had to pull out the previous content first.

> ```{r replace_coldata}

Reviewer: any reason to not make this live?

In this case, we will also log-transform the normalized counts to get a less skewed distribution of expression measures. 
Note that because of the zero counts, the `logNormCounts()` function will add a pseudocount of 1 to each value before computing the log.

```{r sce_normalize}
# Step 1) Group cells with other like cells by clustering.  
qclust <- scran::quickCluster(bladder_sce)

# Step 2) Compute sum factors for each cell cluster grouping.  
bladder_sce <- scran::computeSumFactors(bladder_sce, clusters = qclust)

# Step 3) Normalize using these pooled sum factors and log transform. 
bladder_sce <- scater::logNormCounts(bladder_sce)
```

### Compare normalized data to count data

One way to determine whether our normalization yields biologically relevant results is to plot it and see if similarly labeled samples and cells end up together.

> Because plotting 1000's of genes together isn't practical, we will reduce the dimensions of our data using Principal Components Analysis (PCA). 

Reviewer: ```suggestion
Because plotting 1000's of genes together isn't practical, we will reduce the dimensions of our data using Principal Components Analysis (PCA). 

We will also make the same plot with our unnormalized data, to visualize the effect of normalization on our sample.
Before plotting the unnormalized data, we will log transform the raw counts to make their scaling more comparable to the normalized data. 
To do this we will use the `log1p()` function, which is specifically designed for the case where we want to add 1 to all of our values before taking the log, as we do here.
(We could do something like `log(counts + 1)`, but this is both more efficient and more accurate.)
```
Github is being annoying and not letting me comment this on the next section which is what I wanted, but I think we could add a little more to the explanation of what's happening here to specify the comparison between unnormalized to normalized data in the intro. To me, it was a little confusing when first going through so I can imagine for someone not familiar with this they could also be confused. I added a few lines here that could replace the next paragraph to help with clarity.

We will do this modeling of variance by expression with the `scran::modelGeneVar()` function, saving the results to a new variable. 

```{r model_variance}
gene_variance <- scran::modelGeneVar(bladder_sce)

> ```

Reviewer: another opportunity for a live chunk?

  geom_point(alpha = 0.1) +
  stat_function(fun = metadata(gene_variance)$trend, color = "blue") + 
  xlab("Mean log-expression") + 
  ylab("Variance") +
  theme_bw()

```

Now we can use `scran::getTopHVGs()` to select the genes that have the most biological variation (according to the model) and calculate PCA scores using only those genes. 
Here we are picking the 500 top genes.

```{r get_highvar}
highvar_genes <- scran::getTopHVGs(gene_variance, n = 500)
highvar_pca <- scater::calculatePCA(bladder_sce, subset_row = highvar_genes)
```

```{r highvar_df}
highvar_pca_scores <- data.frame(highvar_pca,
                                 geo_accession = rownames(highvar_pca),
                                 cell_type = bladder_sce$cell_ontology_class)
```

```{r plot_highvar}
ggplot(highvar_pca_scores, aes(x = PC1, y = PC2, color = cell_type)) +
  geom_point() + 
  ggtitle("High variance gene PCA scores") + 
  scale_color_brewer(palette = "Dark2", na.value = "grey70") + 
  theme_bw()
```

Now what we didn't tell you before, but will tell you now, is that the default for `calculatePCA()` is to use the top 500 highest variance genes. 

> This is what we were overriding when we used `ntop = nrow(gbm_sce)`.

Reviewer: ```suggestion
This is what we were overriding when we used `ntop = nrow(bladder_sce)`.
```
